File: apps/udp-iot/packages/logging/ts/src/logger.ts
export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
}

interface LogConfig {
  level: LogLevel;
}

export class Logger {
  private static instances = new Map<string, Logger>();
  private static level = LogLevel.INFO;

  static configure(config: LogConfig): void {
    Logger.level = config.level;
  }

  static getLogger(name: string): Logger {
    let instance = Logger.instances.get(name);
    if (!instance) {
      instance = new Logger(name);
      Logger.instances.set(name, instance);
    }
    return instance;
  }

  private constructor(private readonly name: string) {}

  private log(level: LogLevel, message: string, ...args: unknown[]): void {
    if (level < Logger.level) return;

    const formatted = args.length > 0 ? this.format(message, args) : message;
    const timestamp = new Date().toISOString().replace("T", " ").slice(0, 19);
    const line = `[${timestamp}] [${LogLevel[level]}] [${this.name}] ${formatted}`;

    if (level === LogLevel.ERROR) {
      console.error(line);
    } else {
      console.log(line);
    }
  }

  private format(template: string, args: unknown[]): string {
    const parts: string[] = [];
    let argIdx = 0;
    let start = 0;
    while (true) {
      const brace = template.indexOf("{}", start);
      if (brace === -1) break;
      parts.push(template.slice(start, brace));
      parts.push(argIdx < args.length ? String(args[argIdx++] ?? "") : "{}");
      start = brace + 2;
    }
    parts.push(template.slice(start));
    return parts.join("");
  }

  debug(message: string, ...args: unknown[]): void {
    this.log(LogLevel.DEBUG, message, ...args);
  }

  info(message: string, ...args: unknown[]): void {
    this.log(LogLevel.INFO, message, ...args);
  }

  warn(message: string, ...args: unknown[]): void {
    this.log(LogLevel.WARN, message, ...args);
  }

  error(message: string, ...args: unknown[]): void {
    this.log(LogLevel.ERROR, message, ...args);
  }
}

File: apps/udp-iot/packages/logging/java/src/lib/logging/Logger.java
package lib.logging;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class Logger {
    private static final Map<String, Logger> instances = new ConcurrentHashMap<>();
    private static volatile LogConfig config;
    private static volatile ConsoleHandler console;
    private static volatile FileHandler fileHandler;

    private final String name;

    private Logger(String name) {
        this.name = name;
    }

    public static Logger getLogger(Class<?> clazz) {
        return getLogger(clazz.getSimpleName());
    }

    public static Logger getLogger(String name) {
        return instances.computeIfAbsent(name, Logger::new);
    }

    private static void ensureInitialized() {
        if (config == null) {
            synchronized (Logger.class) {
                if (config == null) {
                    config = LogConfig.load();
                    console = new ConsoleHandler();
                    if (config.hasFile()) {
                        fileHandler = new FileHandler(
                            config.getFilePath(),
                            config.getMaxFileSize(),
                            config.getMaxBackups()
                        );
                    }
                }
            }
        }
    }

    private void log(LogLevel level, String format, Object... args) {
        ensureInitialized();
        if (!level.isEnabled(config.getLevel())) return;

        String message = args.length > 0 ? formatMessage(format, args) : format;
        console.write(level, name, message);
        if (fileHandler != null) {
            fileHandler.write(level, name, message);
        }
    }

    private String formatMessage(String template, Object... args) {
        StringBuilder sb = new StringBuilder();
        int argIdx = 0;
        int start = 0;
        while (true) {
            int brace = template.indexOf("{}", start);
            if (brace == -1) break;
            sb.append(template, start, brace);
            if (argIdx < args.length) {
                Object arg = args[argIdx++];
                sb.append(arg != null ? arg : "null");
            } else {
                sb.append("{}");
            }
            start = brace + 2;
        }
        sb.append(template.substring(start));
        return sb.toString();
    }

    public void debug(String format, Object... args) { log(LogLevel.DEBUG, format, args); }
    public void info(String format, Object... args) { log(LogLevel.INFO, format, args); }
    public void warn(String format, Object... args) { log(LogLevel.WARN, format, args); }
    public void error(String format, Object... args) { log(LogLevel.ERROR, format, args); }
}
